File: java/880_DecodedStringAtIndex.java
class Solution {
    public String decodeAtIndex(String S, int k) {
        long size = 0;
        for (char c : S.toCharArray()) {
            if (Character.isDigit(c)) {
                size *= c - '0';
            } else {
                size++;
            }
        }
        
        int N = S.length();
        for (int i = N - 1; i >= 0; --i) {
            char c = S.charAt(i);
            k %= size;
            if (Character.isLetter(c)) {
                if (k == 0) return "" + c;
                else size--;
            } else {
                size /= c - '0';
            }
        }
        return S.substring(0, 1);
    }
}

File: cpp/880_DecodedStringAtIndex.cpp
class Solution {
public:
    string decodeAtIndex(string S, int K) {
        long int size = 0;
        for (char c : S) {
            int d = c - '0';
            if (d >= 0 && d <= 9) {
                size *= d;
            } else {
                size++;
            }
        }
        
        string res = "";
        for (int i = S.length() - 1; i >= 0; --i) {
            char c = S[i];
            int d = c - '0';
            K %= size;
            if (d >= 0 && d <= 9) {
                size /= d;
            } else {
                if (K == 0) return res + c;
                size--;
            }
        }
        return res + S[0];
    }
};
